contrib/src/main/java/org/apache/apex/malhar/contrib/kudu/AbstractKuduOutputOperator.java

```diff
@@ -0,0 +1,666 @@
+    if ( doNotWriteCols == null) {
+      doNotWriteCols = new HashSet<>();
+    }
+    doNotWriteCols.clear();
+    for (String columnName : allColumnDefs.keySet()) {
+      if ( !(primaryKeyColumnNames.contains(columnName))) {
+        doNotWriteCols.add(columnName);
+      }
+    }
+    kuduExecutionContext.setDoNotWriteColumns(doNotWriteCols);
+    performCommonProcessing(thisDelete,kuduExecutionContext);
+  }
+
+
+  protected void processForInsert(KuduExecutionContext kuduExecutionContext)
+  {
+    Insert thisInsert = kuduTable.newInsert();
+    performCommonProcessing(thisInsert,kuduExecutionContext);
+  }
+
+  @Override
+  public void activate(Context.OperatorContext context)
+  {
+    ApexKuduConnection.ApexKuduConnectionBuilder apexKuduConnectionBuilder = getKuduConnectionConfig();
+    apexKuduConnection = apexKuduConnectionBuilder.build();
+    checkNotNull(apexKuduConnection,"Kudu connection cannot be null");
+    kuduTable = apexKuduConnection.getKuduTable();
+    kuduSession = apexKuduConnection.getKuduSession();
+    kuduClientHandle = apexKuduConnection.getKuduClient();
+    checkNotNull(kuduTable,"Kudu Table cannot be null");
+    checkNotNull(kuduSession, "Kudu session cannot be null");
+    allColumnDefs = new HashMap();
+    primaryKeyColumnNames = new HashSet<>();
+    kuduColumnBasedGetters = new HashMap();
+    buildGettersForPojoPayload();
+    isInSafeMode = false;
+    isInReconcilingMode = false;
+    reconcilingWindowId = Stateless.WINDOW_ID;
+    if ( (context.getValue(Context.OperatorContext.ACTIVATION_WINDOW_ID) != Stateless.WINDOW_ID) &&
+        context.getValue(Context.OperatorContext.ACTIVATION_WINDOW_ID) <
+        windowDataManager.getLargestCompletedWindow()) {
+      isInSafeMode = true;
+      reconcilingWindowId = windowDataManager.getLargestCompletedWindow() + 1;
+      isInReconcilingMode = false;
+    }
+  }
+
+  @Override
+  public void deactivate()
+  {
+    try {
+      apexKuduConnection.close();
+    } catch (Exception e) {
+      LOG.error("Could not close kudu session and resources because " + e.getMessage(), e);
+    }
+  }
+
+  @Override
+  public void beforeCheckpoint(long l)
+  {
+    // Nothing to be done here. Child classes can use this if required
+  }
+
+  @Override
+  public void checkpointed(long l)
+  {
+    // Nothing to be done here. Child classes can use this if required
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+    try {
+      windowDataManager.committed(windowId);
+    } catch (IOException e) {
+      LOG.error("Error while committing the window id " + windowId + " because " + e.getMessage());
+      throw new RuntimeException(e.getMessage());
+    }
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    windowDataManager = getWindowDataManager();
+    if ( windowDataManager == null) {
+      windowDataManager = new FSWindowDataManager();
```

Comment:
I think FSWindowManager makes a sense default for the following reason:

We are actually using the values stored in the WindowManager to implement logic to identify states related
1. A set of tuples are being replayed from the upstream buffer server
2. Identifying a reconcile window ( The first window that is resuming since the last exit of the window ) 

In the reconciling window phase, we are transferring the onus of "EXACTLY_ONCE" semantics to the end user ( please see below comments for this aspect ) so that the business logic can be used to decide whether we want to mutate the kudu store given the above two situations.